### PR TITLE
Bids 3143/user settings page

### PR DESCRIPTION
**Pull request by D13ce**
This PR
* Adds the `user/settings` page
* Adds `UserSettingsSubscriptions.vue`, `UserSettingsEmail.vue`, `UserSettingsPassword.vue` and `UserSettingsDeleteAccount.vue` for `/user/settings`
* Adds `yup` for form validation via `vee-validate`
* Improves `BcDialogDelete.vue` functionality: It now has a `severity` prop and if set to `'danger'`, the styling is changed accordingly

### CODE REVIEW

**Review by MauserBitfly — COMMENTED** [review]
Nicely done, left some comments.

File: frontend/components/bc/header/MainHeader.vue [hunk 1]
```diff
@@ -46,8 +46,16 @@ const toggleMegaMenu = (evt: Event) => {
 
 const isMobileMegaMenuOpen = computed(() => megaMenu.value?.isMobileMenuOpen)
 
+const navigateToUserSettings = async () => {
+  await navigateTo('../user/settings')
+}
+
 const userMenu = computed(() => {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/bc/header/MainHeader.vue:53]: Very good catch, thanks. Fixed.

File: frontend/assets/css/colors.scss [hunk 2]
```diff
@@ -55,8 +56,11 @@
   --button-color-pressed: var(--primary-orange-pressed);
   --button-color-disabled: var(--grey);
   --button-color-dark-pattern: var(--light-grey-2);
+  --button-color-dangerous: var(--flashy-red);
+  --button-color-dangerous-hover: var(--dark-red);
   --button-text-color-dark-pattern: var(--light-black);
   --button-text-color-disabled: var(--grey-4);
+  --button-text-color-dangerous: var(--grey-4);
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/assets/css/colors.scss:63]: You are right, fixed.

File: frontend/components/user/settings/UserSettingsSubscriptions.vue [hunk 3]
```diff
@@ -0,0 +1,77 @@
+<script lang="ts" setup>
+const { t: $t } = useI18n()
+const { currentPremiumSubscription } = useProductsStore()
+const { stripeCustomerPortal, isStripeDisabled } = useStripe()
+
+const buttonsDisabled = defineModel<boolean | undefined>({ required: true })
+
+async function buttonCallback () {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/user/settings/UserSettingsSubscriptions.vue]: Done.

File: frontend/pages/user/settings.vue [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+await useAsyncData('get_products', () => getProducts())
+watch(products, () => {
+  if (products.value?.stripe_public_key) {
+    stripeInit(products.value.stripe_public_key)
+  }
+}, { immediate: true })
+
+if (!isLoggedIn.value) {
+  // only users that are logged in can view this page
+  await navigateTo('/login')
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/pages/user/settings.vue:20]: Added Todo.

File: frontend/components/bc/header/MainHeader.vue [hunk 5]
```diff
@@ -46,8 +46,16 @@ const toggleMegaMenu = (evt: Event) => {
...
   return [
+    {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/bc/header/MainHeader.vue:51]: As discussed, we will always show this but we will forward to V1 via `ShowInDevelopment` until the backend is fully implemented.

File: frontend/assets/css/main.scss [hunk 6]
```diff
@@ -49,6 +49,24 @@ a {
   @include container;
 }
 
+@mixin button-dangerous {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/assets/css/main.scss:52]: I like that. Done.

File: frontend/components/bc/dialog/BcDialogDelete.vue [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+interface Props {
+  title?: string
+  warning?: string
+  noLabel?: string // defaults to "Cancel"
+  yesLabel?: string // defaults to "Delete"
+}
+const { props, dialogRef } = useBcDialog<Props>({ showHeader: false })
+
+const closeDialog = (response: boolean) => {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/bc/dialog/BcDialogDelete.vue]: As discussed, I thought about that but chose not to do it as there were many (minor) styling differences. But, as you've suggested, I have talked with the designer and those differences are now gone. I have deleted `BcDialogDelete` and everything now uses an updated `BcDialogConfirm` :bow:

File: frontend/components/dashboard/DashboardControls.vue [hunk 8]
```diff
@@ -119,17 +119,26 @@ const deleteButtonOptions = computed(() => {
 })
 
 const onDelete = () => {
-  const languageKey = deleteButtonOptions.value.deleteDashboard ? 'dashboard.deletion.delete_text' : 'dashboard.deletion.clear_text'
-
-  dialog.open(BcDialogConfirm, {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/dashboard/DashboardControls.vue:124]: Of course. Done.

File: frontend/components/user/settings/UserSettingsEmail.vue [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+      <InputText
+        id="confirm-email"
+        v-model="confirmEmail"
+        :class="{ 'p-invalid': errors?.confirmEmail }"
+        aria-describedby="text-error"
+        @focusin="confirmEmailError = undefined"
+        @focusout="confirmEmailError = errors?.confirmEmail"
+      />
+      <div class="p-error">
+        {{ confirmEmailError || '&nbsp;' }}
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/user/settings/UserSettingsEmail.vue]: Okay, this required quite the rework but I believe you will like both the implementation and the result better.

File: frontend/components/user/settings/UserSettingsEmail.vue [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+      {{ $t('user_settings.email.password') }}
+    </label>
+    <div class="input-row">
+      <InputText
+        id="password"
+        v-model="password"
+        type="password"
+        :class="{ 'p-invalid': errors?.password }"
+        aria-describedby="text-error"
+        @focusin="passwordError = undefined"
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/user/settings/UserSettingsEmail.vue]: Did something else: I have introduced the `yup` package which is used for validation now.

File: frontend/components/dashboard/DashboardControls.vue (hunk 8 above)
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply frontend/components/dashboard/DashboardControls.vue:124]: Nice thanks.

File: frontend/components/user/settings/UserSettingsPassword.vue [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+import { API_PATH } from '~/types/customFetch'
+
+const { t: $t } = useI18n()
+const { fetch } = useCustomFetch()
+const toast = useBcToast()
+
+const { handleSubmit, errors, defineField } = useForm({
+  validationSchema: yupObject({
+    oldPassword: passwordValidation($t),
+    newPassword: passwordValidation($t),
```
Comment (MauserBitfly) [inline frontend/components/user/settings/UserSettingsPassword.vue:13]: Please add a Validator for the new password that it does not match the ole one

File: frontend/components/user/settings/UserSettingsEmail.vue [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  buttonsDisabled.value = true
+  try {
+    await fetch(API_PATH.USER_CHANGE_EMAIL, {
+      body: {
+        password: values.password,
+        email: values.newEmail
+      }
+    })
```
Comment (MauserBitfly) [inline frontend/components/user/settings/UserSettingsEmail.vue:36]: maybe a success toast would be nice ... the user does not get any feedback right now (if success)

File: frontend/components/user/settings/UserSettingsPassword.vue [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    return
+  }
+
+  buttonsDisabled.value = true
+  try {
+    await fetch(API_PATH.USER_CHANGE_PASSWORD, {
+      body: {
+        password: values.newPassword
+      }
+    })
```
Comment (MauserBitfly) [inline frontend/components/user/settings/UserSettingsPassword.vue:35]: As for the email, please give the user a positive feedback on success

**Review by MauserBitfly — COMMENTED** [review]
Nice Job, I love the new Yup validation stuff and that you merged the two Dialogs.

Left some more comments ;-)

File: frontend/components/bc/header/MainHeader.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/bc/header/MainHeader.vue:49]: This is based on the design: 
image
If you want, I can send you a link in DMs.

[the remark this answers is not in the document]
Comment (MauserBitfly) [reply frontend/components/bc/header/MainHeader.vue:49]: I still think it's wrong ... but ok

File: frontend/components/user/settings/UserSettingsPassword.vue (hunk 11 above)
Comment (D13ce) [reply frontend/components/user/settings/UserSettingsPassword.vue]: I like the idea, done.

File: frontend/components/user/settings/UserSettingsEmail.vue (hunk 12 above)
Comment (D13ce) [reply frontend/components/user/settings/UserSettingsEmail.vue:36]: Good idea, added. Furthermore the form now gets reset on success.

File: frontend/components/user/settings/UserSettingsPassword.vue (hunk 13 above)
Comment (D13ce) [reply frontend/components/user/settings/UserSettingsPassword.vue:35]: Good idea, added. Furthermore the form now gets reset on success.

File: frontend/components/bc/dialog/BcDialogConfirm.vue [hunk 14]
```diff
@@ -43,15 +51,24 @@ const closeDialog = (response: boolean) => {
...
 
   .footer {
     display: flex;
-    justify-content: center;
+    align-items: center;
+    justify-content: flex-end;
     margin-top: var(--padding);
     gap: var(--padding);
+
+    &.danger>:first-child {
```
[the remark this answers is not in the document]
Comment (D13ce) [reply frontend/components/bc/dialog/BcDialogConfirm.vue:66]: Jeah, this must be the remnant of some re-factoring. Cleaned it up a but.
I have skipped the bonus points though :grimacing:

[the remark this answers is not in the document]
Comment (MauserBitfly) [reply frontend/components/bc/dialog/BcDialogConfirm.vue]: you get an A (no + ) for that assignment ;-)

File: frontend/components/user/settings/UserSettingsEmail.vue (hunk 12 above)
Comment (MauserBitfly) [reply frontend/components/user/settings/UserSettingsEmail.vue:36]: get an A+ for this one dough

File: frontend/components/user/settings/UserSettingsPassword.vue (hunk 13 above)
Comment (MauserBitfly) [reply frontend/components/user/settings/UserSettingsPassword.vue:35]: same - A+

**Review by MauserBitfly — APPROVED** [review]
LGTM

**Pull request merged**